### PR TITLE
KEYCLOAK-13983 Include algorithm parameters

### PR DESCRIPTION
**Pull request by Norwae**
Adds the capacity to add both public and secret algorithm
specific data to a PasswordCredentialModel. The default
way to create the models in maintained to minimize the change
impact for the default hash infrastructure.

Publishes the PasswordCredentialModel constructor to
ease in building such extended credential models.

### CODE REVIEW

**Review by pedroigor — CHANGES_REQUESTED** [review]
LGTM. But we also need changes to `CredentialEntity` to persist the algorithm data.

The property could have a more generic name so that you can use not only for the algorithm but to persist any other information you want to attach to the credential.

@mposolda Wdyt?

File: server-spi/src/main/java/org/keycloak/models/credential/PasswordCredentialModel.java
Comment (mposolda) [inline server-spi/src/main/java/org/keycloak/models/credential/PasswordCredentialModel.java:18]: Just for the sake of compatibility with other implementations, could we keep the constructor private and instead introduce other static method like for example:

```
public static PasswordCredentialModel createFromValues(String algorithm, byte[] salt, int hashIterations, String 
encodedPassword, MultivaluedHashMap<String, String> otherCredentialData,, MultivaluedHashMap<String, String> otherSecretData)
```
The current method signature "createFromValues" needs to be kept as well for the backwards compatibility, but it can just call the new method. Something like:
```
public static PasswordCredentialModel createFromValues(String algorithm, byte[] salt, int hashIterations, String encodedPassword) {
    createFromValues(algorithm, salt, hashIterations, encodedPassword, Collections.emptyMap(), Collections.emptyMap());
}
```

**Review by mposolda — CHANGES_REQUESTED** [review]
Thanks for the PR! Few points:

1) +1 to what @pedroigor mentioned about have more generic name. Maybe something like "otherData" will work?

2) Could we have generic abstract class like for example:
```
/**
 * Useful for the credentialData or secretData for the credentials, which may contain some "generic" 
metadata for the particular credential. For example advanced password hashing algorithms like SCrypt 
needs to contain various data specific to SCrypt algorithm
 */
public abstract class GenericData {
    
   /
    private MultivaluedHashMap<String, String> otherData;
   // TODO: Getters/setters...
}
````

Then the PasswordCredentialData and PasswordSecretData may just inherit from this class?

3) Note I used MutlivaluedHashMap instead of "simple" Map to make sure that values of some keys can be eventually multivalued. When we're doing more generic solution, my vote would be to add multivalued from the beginning to not see similar limitations, which you're seeing now :-)

4) @pedroigor I don't think that there is a need to change CredentialEntity because new metadata are directly part of the JSON? So regarding DB representation, things are not changed. 

5) We should make sure that JSON migrated from previous versions will still work. And those JSONs obviously don't have "otherData" yet in them. In other words, it should be still possible to parse JSON like 
```
{
   "hashIterations": 10,
   "algorithm": "foo"
}
```
without seeing some strange errors. It will be good if you can add automated test for this into CredentiaModelTest. We will later need to run the migrations test in the pipeline to doublecheck there are no regressions.

File: server-spi/src/main/java/org/keycloak/models/credential/PasswordCredentialModel.java
Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/models/credential/PasswordCredentialModel.java:18]: @mposolda Yeah, you're right about the credential data. I was thinking about a separated and more general-purpose thing like what you mentioned before in regards to something like "otherClaims". And possibly avoid migration, etc.

But if you think we can reuse that column, that is fine.

Comment (Norwae) [reply server-spi/src/main/java/org/keycloak/models/credential/PasswordCredentialModel.java:18]: I agree, we can move to a static factory + private constructor.

File: server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordSecretData.java
Comment (pedroigor) [inline server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordSecretData.java:51]: Could you lazily create the instance on the getter method and avoid creating this instance every time an instance is created? We are working on reducing footprint and small changes like will help.

File: themes/src/main/resources/theme/keycloak/common/resources/node_modules/rcue/dist/img/git-logo.svg
Comment (pedroigor) [inline themes/src/main/resources/theme/keycloak/common/resources/node_modules/rcue/dist/img/git-logo.svg:1]: Can we remove this change from this PR? It looks OK, but I would prefer making theme related changes in a separate PR. Otherwise, we are going to need to include a few more people to the review :)

**Review by pedroigor — CHANGES_REQUESTED** [review]
LGTM. I've only these last comments and I think we're ready to go.

File: themes/src/main/resources/theme/keycloak/common/resources/node_modules/rcue/dist/img/git-logo.svg
Comment (Norwae) [reply themes/src/main/resources/theme/keycloak/common/resources/node_modules/rcue/dist/img/git-logo.svg]: That was actually a mistake - I have restored it. Thanks for the catch. 👍

File: server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordSecretData.java
Comment (Norwae) [reply server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordSecretData.java:51]: Doing so would require me to give up the final modified on the fields (lazy creation), or alternatively allow the field to be null. Since all the explicitly-modeled fields are final in these types, I would prefer to not break this pattern (principle of least surprise). 

Again, I will defer to your judgement here, but I at least wanted to raise the point.

Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/models/credential/dto/PasswordSecretData.java:51]: I prefer the lazy creation or maybe you can just initialize as an empty collection when the instance is created using this constructor?

**Review by pedroigor — APPROVED** [review]
@Norwae Thanks. 

@mposolda LGTM, let us know if there is something else.

**Review by mposolda — APPROVED** [review]
@Norwae @pedroigor Sorry for late review. Finally approving and merging now.

**Pull request merged**
